c_stubs/stub_tap_open.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Citrix 
+ *
+ * This could be replaced by https://github.com/mirage/ocaml-tuntap
+ * if more features are required.
+ */
+
+#include <string.h>
+#include <unistd.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <linux/if_tun.h>
+#include <fcntl.h>
+
+
+#include <caml/mlvalues.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+#include <caml/memory.h>
+
+#define PATH_NET_TUN "/dev/net/tun"
+
+/* open /dev/net/tun and set it up
+ * external _tap_open : string -> Unix.file_descr = "stub_tap_open"
+ */
+CAMLprim value stub_tap_open(value ocaml_ifname)
+{
+    CAMLparam1(ocaml_ifname);
+    unsigned int features;
+    struct ifreq ifr;
+    char *ifname = String_val(ocaml_ifname);
+
+    memset(&ifr, 0, sizeof(ifr));
+    ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
+
+    size_t len = strlen(ifname);
+    if (len == 0) {
+        caml_failwith("empty string argument in " __FILE__);
+    }
+    if (len >= IFNAMSIZ) {
+        caml_failwith("string argument too long in "__FILE__);
+    }
+    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
+
+    int fd = open(PATH_NET_TUN, O_RDWR);
+    if (fd < 0) {
+        caml_failwith("open(" PATH_NET_TUN ") failed in " __FILE__);
+    }
+
+    if (ioctl(fd, TUNGETFEATURES, &features) == -1) {
+        close(fd);
+        caml_failwith("TUNGETFEATURES failed in " __FILE__);
+    }
+
+    if (features & IFF_ONE_QUEUE) {
+        ifr.ifr_flags |= IFF_ONE_QUEUE;
+    }
+
+    int ret = ioctl(fd, TUNSETIFF, (void *) &ifr);
+    if (ret != 0) {
```

Comment:
This is more a curiosity. Why not inline as you did above? We don't use `ret` for anything else.